File: dev-portal/python/postman_sdk/types/collection.py
"""
Collection-related types
"""

from typing import Any
from pydantic import BaseModel, Field


class CollectionFork(BaseModel):
    """Collection fork info"""
    label: str
    created_at: str = Field(alias="createdAt")
    source: str = Field(alias="from")

    class Config:
        populate_by_name = True


class Collection(BaseModel):
    """Collection summary (from list endpoint)"""
    id: str
    uid: str
    name: str
    owner: str | None = None
    created_at: str | None = Field(None, alias="createdAt")
    updated_at: str | None = Field(None, alias="updatedAt")
    fork: CollectionFork | None = None

    class Config:
        populate_by_name = True


class CollectionInfo(BaseModel):
    """Collection info block"""
    name: str
    description: str | None = None
    schema_url: str = Field(alias="schema")
    postman_id: str | None = Field(None, alias="_postman_id")

    class Config:
        populate_by_name = True


class CollectionDetails(Collection):
    """Full collection details"""
    info: CollectionInfo | None = None
    item: list[dict[str, Any]] = Field(default_factory=list)
    auth: dict[str, Any] | None = None
    variable: list[dict[str, Any]] = Field(default_factory=list)
    event: list[dict[str, Any]] = Field(default_factory=list)


class ForkResult(BaseModel):
    """Fork collection result"""
    success: bool
    collection: Collection | None = None
    error: str | None = None


class CollectionMapping(BaseModel):
    """Collection mapping (source to target)"""
    source_uid: str
    target_uid: str
    name: str
    mock_url: str | None = None

File: dev-portal/typescript/src/types/collection.ts
/**
 * Collection-related types
 */

/**
 * Collection summary (from list endpoint)
 */
export interface Collection {
  id: string;
  uid: string;
  name: string;
  owner?: string;
  createdAt?: string;
  updatedAt?: string;
  fork?: {
    label: string;
    createdAt: string;
    from: string;
  };
}

/**
 * Full collection details
 */
export interface CollectionDetails extends Collection {
  info: {
    name: string;
    description?: string;
    schema: string;
    _postman_id?: string;
  };
  item: CollectionItem[];
  auth?: CollectionAuth;
  variable?: CollectionVariable[];
  event?: CollectionEvent[];
}

/**
 * Collection item (folder or request)
 */
export interface CollectionItem {
  id?: string;
  name: string;
  description?: string;
  item?: CollectionItem[];
  request?: CollectionRequest;
  response?: CollectionResponse[];
  event?: CollectionEvent[];
}

/**
 * Collection request
 */
export interface CollectionRequest {
  method: string;
  url: string | CollectionUrl;
  header?: CollectionHeader[];
  body?: CollectionBody;
  auth?: CollectionAuth;
  description?: string;
}

/**
 * Collection URL
 */
export interface CollectionUrl {
  raw: string;
  protocol?: string;
  host?: string[];
  path?: string[];
  query?: Array<{ key: string; value: string; disabled?: boolean }>;
  variable?: Array<{ key: string; value: string }>;
}

/**
 * Collection header
 */
export interface CollectionHeader {
  key: string;
  value: string;
  description?: string;
  disabled?: boolean;
}

/**
 * Collection body
 */
export interface CollectionBody {
  mode: 'raw' | 'formdata' | 'urlencoded' | 'file' | 'graphql';
  raw?: string;
  formdata?: Array<{ key: string; value: string; type?: string }>;
  urlencoded?: Array<{ key: string; value: string }>;
  options?: {
    raw?: {
      language?: string;
    };
  };
}

/**
 * Collection auth
 */
export interface CollectionAuth {
  type: string;
  bearer?: Array<{ key: string; value: string }>;
  basic?: Array<{ key: string; value: string }>;
  apikey?: Array<{ key: string; value: string }>;
}

/**
 * Collection variable
 */
export interface CollectionVariable {
  key: string;
  value: string;
  type?: string;
  disabled?: boolean;
}

/**
 * Collection event (pre-request/test scripts)
 */
export interface CollectionEvent {
  listen: 'prerequest' | 'test';
  script: {
    type: string;
    exec: string[];
  };
}

/**
 * Collection response (saved example)
 */
export interface CollectionResponse {
  id?: string;
  name: string;
  originalRequest?: CollectionRequest;
  status?: string;
  code?: number;
  header?: CollectionHeader[];
  body?: string;
}

/**
 * Fork collection request
 */
export interface ForkCollectionRequest {
  collectionUid: string;
  label: string;
  targetWorkspaceId: string;
}

/**
 * Fork collection result
 */
export interface ForkResult {
  success: boolean;
  collection?: Collection;
  error?: string;
}

/**
 * Collection mapping (source to target)
 */
export interface CollectionMapping {
  sourceUid: string;
  targetUid: string;
  name: string;
  mockUrl?: string;
}
